### PR TITLE
Krylov as an integrator

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
Add `krylov` method as an `Integrator` for `SeSolver` porting `krylovsolve` to v5 format.

The state evolution code was rewritten to use the qutip's data object instead of numpy arrays, but is otherwise the same.

Since an integrator can be used without knowing the length of the evolution, some changes where as to when the krylov matrices are recomputed. Instead of choosing the step length so the error of the whole evolution is under `atol`, it is chosen so each steps stays under it and an options is available to compute it only once or at each step.

A default `krylov_dim` was needed for the integrator, so I set it the scale as the square root of the size of the system. This is not ideal, but seems to work fine. When using it through `krylovsolve`, `krylov_dim`is a required parameter.

### CODE REVIEW

**Comment by hodgestar** [comment]
@Ericgig Could you merge in dev.major and resolve conflicts?

File: qutip/solver/ode/krylov.py
Comment (hodgestar) [inline qutip/solver/ode/krylov.py:44]: Where does the approximate expression for err come from? I would expect it to depend on the structure of the eigenvalues of the system, and not just the dimension.

File: qutip/solver/ode/krylov.py
Comment (hodgestar) [inline qutip/solver/ode/krylov.py:218]:
```suggestion
            Dimension of Krylov approximation subspaces used for the time
            evolution approximation. If the defaut 0 is given, the dimension is calculated
            from the system size N, using `min(int((N + 100)**0.5), N-1)`.
```

**Review by hodgestar — CHANGES_REQUESTED** [review]
It's great that we can implement this as an integrator now. I left some suggestions, but it looks good overall.

File: qutip/solver/ode/krylov.py
Comment (Ericgig) [reply qutip/solver/ode/krylov.py:97]: The `N` is usually quite small and we need to compute it's eigen decomposition. The sparse eigen solver is a lot slower and less precise..

File: qutip/solver/ode/krylov.py
Comment (Ericgig) [reply qutip/solver/ode/krylov.py:150]: The `root_scalar` call need to look for a `0` and converge better with the log.

File: qutip/solver/ode/krylov.py
Comment (Ericgig) [reply qutip/solver/ode/krylov.py:44]: I tried it with various Hamiltonian and states. The eigenvalues and state affect the constant and power of dt.
However, larger dims decrease the error and `dt` increase it. 
`_compute_max_step` already compute `dt` as a function of `err` for fixed `krylov_dims`. We could compute `krylov_dims` with fixed `dt` and `err` or find optimal `krylov_solve` and `dt`with fixed. But I am not sure it's worth it so I just left a comment that it could be done.

File: qutip/solver/ode/krylov.py
Comment (hodgestar) [reply qutip/solver/ode/krylov.py:149]: Ah, I see. I will add a comment about it.

File: qutip/solver/ode/krylov.py
Comment (hodgestar) [inline qutip/solver/ode/krylov.py:134]:
```suggestion
            # we divide by atol and take the log so that the error returned is 0
            # at atol, which is convenient for calling root_scalar with.
            return np.log(_data.norm.l2(
```

**Review by hodgestar — APPROVED** [review]

**Comment by hodgestar** [comment]
Thanks for addressing all the issues so quickly. I added a commit for the krylov_error comment I wanted. Happy for this to be merged if the latest test run passes.

**Comment by Ericgig** [comment]
I am working on a merge master into dev.major PR and there will be conflict with the krylov documentation. It would be easier to wait for that and fix the conflict here. Otherwise it will be lost in a 60 files PR.

**Comment by hodgestar** [comment]
Happy for you to merge this once you're done with the merge of master.

**Comment by Ericgig** [comment]
I updated the krylov page after the merger. The code was not touched since the review.

**Review by hodgestar — APPROVED** [review]

**Pull request merged**
